Scripts/Runtime/Jobs/Collections/CollectionAccessControl.cs

```diff
@@ -15,80 +16,117 @@ namespace Anvil.Unity.DOTS.Jobs
+        // INTERNAL INTERFACES
+        //*************************************************************************************************************
 
-        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.SubsystemRegistration)]
-        private static void Init()
+        internal interface ICollectionAccessControl<TKey> : ICollectionAccessControl
         {
-            Dispose();
+            CollectionAccessController<TKey> GetOrCreate(TKey key = default);
+            void Remove(CollectionAccessController<TKey> collectionAccessController);
         }
-        
-        /// <summary>
-        /// Returns an instance of an <see cref="CollectionAccessController{TContext, TKey}"/> for a given key.
-        /// Will create a new one if it doesn't already exist.
-        /// </summary>
-        /// <param name="key">The key to lookup or create the instance of the <see cref="CollectionAccessController{TContext, TKey}"/>.</param>
-        /// <returns>The <see cref="CollectionAccessController{TContext,TKey}"/> instance.</returns>
-        public static CollectionAccessController<TContext, TKey> GetOrCreate(TKey key = default)
+
+        internal interface ICollectionAccessControl
+        {
+            void RemoveAndDisposeAll();
+        }
+
+        //*************************************************************************************************************
+        // INTERNAL HELPER
+        //*************************************************************************************************************
+
+        private class CollectionAccessControlImpl<TContext, TKey> : ICollectionAccessControl<TKey>
         {
-            if (!s_Controllers.TryGetValue(key, out CollectionAccessController<TContext, TKey> accessController))
+            private readonly Dictionary<TKey, CollectionAccessController<TKey>> m_Controllers = new Dictionary<TKey, CollectionAccessController<TKey>>();
+
+            private bool m_IsRemovingAndDisposingAll;
+
+            internal CollectionAccessControlImpl()
             {
-                accessController = new CollectionAccessController<TContext, TKey>(key);
-                accessController.OnDisposed += CollectionAccessController_OnDisposed;
-                s_Controllers.Add(key, accessController);
             }
 
-            return accessController;
+            public CollectionAccessController<TKey> GetOrCreate(TKey key = default)
+            {
+                if (!m_Controllers.TryGetValue(key, out CollectionAccessController<TKey> collectionAccessController))
+                {
+                    collectionAccessController = new CollectionAccessController<TKey>(this, key);
+                    m_Controllers.Add(key, collectionAccessController);
+                }
+
+                return collectionAccessController;
+            }
+
+            public void Remove(CollectionAccessController<TKey> collectionAccessController)
```

Comment:
Should this be internal?

```suggestion
            internal void Remove(CollectionAccessController<TKey> collectionAccessController)
```

Reply:
It has to be public because of the interface. 

But the interfaces are internal and the class itself is private so no one will accidentally get here.